smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -0,0 +1,654 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import json
+import logging
+import os
+import pathlib
+import signal
+import sys
+import threading
+import time
+import typing as t
+
+from dataclasses import dataclass, field
+from types import FrameType
+
+from watchdog.observers import Observer
+from watchdog.observers.api import BaseObserver
+from watchdog.events import PatternMatchingEventHandler, LoggingEventHandler
+from watchdog.events import FileCreatedEvent, FileModifiedEvent
+
+from smartsim._core.control.job import Job, JobEntity
+from smartsim._core.control.jobmanager import JobManager
+from smartsim._core.launcher.stepInfo import StepInfo
+from smartsim._core.launcher.cobalt.cobaltLauncher import CobaltLauncher
+
+from smartsim._core.launcher.launcher import Launcher
+from smartsim._core.launcher.local.local import LocalLauncher
+from smartsim._core.launcher.lsf.lsfLauncher import LSFLauncher
+from smartsim._core.launcher.pbs.pbsLauncher import PBSLauncher
+from smartsim._core.launcher.slurm.slurmLauncher import SlurmLauncher
+from smartsim._core.utils.helpers import get_ts
+from smartsim._core.utils.serialize import TELMON_SUBDIR, MANIFEST_FILENAME
+
+from smartsim.error.errors import SmartSimError
+from smartsim.status import TERMINAL_STATUSES
+
+
+"""
+Telemetry Monitor entrypoint
+"""
+
+# kill is not catchable
+SIGNALS = [signal.SIGINT, signal.SIGQUIT, signal.SIGTERM, signal.SIGABRT]
+_EventClass = t.Literal["start", "stop", "timestep"]
+_ManifestKey = t.Literal["timestamp", "model", "orchestrator", "ensemble", "run_id"]
+_JobKey = t.Tuple[str, str]
+
+
+@dataclass
+class Run:
+    """Model containing entities of an individual start call for an experiment"""
+
+    timestamp: int
+    models: t.List[JobEntity]
+    orchestrators: t.List[JobEntity]
+    ensembles: t.List[JobEntity]
+
+    def flatten(
+        self, filter_fn: t.Optional[t.Callable[[JobEntity], bool]] = None
+    ) -> t.List[JobEntity]:
+        """Flatten runs into a list of SmartSimEntity run events"""
+        entities = self.models + self.orchestrators + self.ensembles
+        if filter_fn:
+            entities = [entity for entity in entities if filter_fn(entity)]
+        return entities
+
+
+@dataclass
+class RuntimeManifest:
+    """The runtime manifest holds meta information about the experiment entities created
+    at runtime to satisfy the experiment requirements."""
+
+    name: str
+    path: pathlib.Path
+    launcher: str
+    runs: t.List[Run] = field(default_factory=list)
+
+
+def hydrate_persistable(
+    entity_type: str,
+    persistable_entity: t.Dict[str, t.Any],
+    exp_dir: pathlib.Path,
+) -> t.List[JobEntity]:
+    """Map entity data persisted in a manifest file to an object"""
+    entities: t.List[JobEntity] = []
+
+    if "shards" in persistable_entity or "models" in persistable_entity:
+        container = "shards" if "shards" in persistable_entity else "models"
```

Comment:
used set intersection to clarify the key lookup:

```
    db_keys = {"shards", "models"}
    db_keys = db_keys.intersection(persistable_entity.keys())
    if db_keys:
```

added `_hydrate_persistable` to avoid the duplication:
```
    db_keys = {"shards", "models"}
    db_keys = db_keys.intersection(persistable_entity.keys())
    if db_keys:
        container = "shards" if "shards" in db_keys else "models"
        for shard in persistable_entity[container]:            
            entity = _hydrate_persistable(shard, entity_type, str(exp_dir))
            entities.append(entity)

        return entities

    entity = _hydrate_persistable(persistable_entity, entity_type, str(exp_dir))
    entities.append(entity)
    entities.append(entity)
```

Reply:
I think using the set should clean it up... Since the events don't really use anything other than the Job metadata, I'd like to avoid adding any more complexity to the abstraction as long as possible.